### DISCUSSION TITLE
feedback problem

### DISCUSSION THREAD

**Discussion by mary-9555** [jspsych/jsPsych#2195] [Q&A] [answered]
Hi there.
Can someone please tell me what is wrong with this coding? or what is missing in it? I would really appreciate it.
```js
var image_trial = {
  type: 'image-button-response',
  stimulus: 'img/b/004_o_m_s_b.jpeg',
  choices: ['sad', 'frightened', 'neutral','angry','disgusting','happy'],
  data: {
    stimulus_type: 'image'
  },
  on_finish: function(data){
    if(jsPsych.pluginAPI.compareKeys(data.response, "sad")){
      data.correct = true;
    } else {
      data.correct = false; 
    }
  }
}

var feedback = {
  type: 'instructions',
  stimulus: function(){
    var last_trial_correct = jsPsych.data.get().last(1).values()[0].correct;
    if(last_trial_correct){
      return "<p>Correct!</p>"; 
    } else {
      return "<p>Wrong!... The correct answer was 'sad'.</p>";
      show_clickable_nav; true,
    button_label_previous; 'previous',
    button_label_next; '<h2>next</h2>'
    }
  }
}

timeline.push(image_trial, feedback);
```

**Comment by becky-gilbert** [comment, marked as answer]
Hi @mary-9555, can you say more about what's happening when you run the code? Are you seeing any errors in the browser's JavaScript console?

I did notice a few things:
1. The `jsPsych.pluginAPI.compareKeys` function is just meant for comparing _keyboard_ responses, not button responses. So in your `image_trial`, you should compare the participant's response and correct response directly in the `on_finish` function. The participant's button response is stored using the indices in the `choices` array, and in your case "sad" is the first button choice (index = 0), so you want to check if the response was "sad" like this: `if (data.response == 0) { ...`
2. In your `feedback` trial, using the `instructions` plugin might not be ideal here, because that plugin is meant for displaying several pages that the participant can go forward/backwards through. If you just need to display a single page with the "correct" or "incorrect" feedback, then you might want to use one of the `html-*` plugins. If you want the participant to have to click a button to continue on from the feedback page, you can do that using the `html-button-response` plugin.

Also, if you do want to stick with the `instructions` plugin for your feedback trial, there are some problems with the way this trial is defined:

3. Instead of a `stimulus` parameter, the `instructions` plugin has a `pages` parameter that must be an _array_ of one or more strings (the HTML-formatted content for each page). See the [instructions plugin documentation](https://www.jspsych.org/6.3/plugins/jspsych-instructions/) for more info.
4. You have a misplaced curly bracket in this trial. You should end the if/else statement and `stimulus` function with two curly brackets after the last `return` statement, and the remaining 3 trial parameters should be on the same level as `type` and `pages`. 
5. The trial parameters and values should be separated by a colon (:) rather than a semi-colon (;). 
```js
var feedback = {
  type: 'instructions',
  pages: function(){
    var last_trial_correct = jsPsych.data.get().last(1).values()[0].correct;
    if(last_trial_correct){
      return ["<p>Correct!</p>"]; 
    } else {
      return ["<p>Wrong!... The correct answer was 'sad'.</p>"];
    }
  },
  show_clickable_nav: true,
  button_label_previous: 'previous',
  button_label_next: '<h2>next</h2>'
}
```

**Reply by mary-9555** [reply]
Thanks for your reply Becky.
You are right, this is the problem: The `jsPsych.pluginAPI.compareKeys` function is just meant for comparing **keyboard responses,** not **_button_** responses!

I followed this recommendation: in your image_trial, you should compare the participant's response and correct response directly in the on_finish function: if (data.response == "sad") { ...

But it didn't work !!
There was a wrong! feedback for every answer! (even when you choose the correct button!).
I wish there was a `jsPsych.pluginAPI.comparebuttons` function for this kinda task...

**Reply by becky-gilbert** [reply]
Hi @mary-9555, oh sorry, I just remembered that the button responses are stored as _numbers_ in the trial data, where 0 refers to the first button option in `choices`, 1 is the second button, etc. So my response above is wrong - I'll edit it so that anyone else reading this doesn't get confused.

The "sad" button is first in your `choices` array (i.e. index 0), so in the `on_finish` function for your `image_trial`, you want to check whether `data.response` is 0, like this:
```js
on_finish: function(data) {
  if (data.response == 0) {  // 0 is 'sad', 1 is 'frightened', 2 is 'neutral', etc.
    data.correct = true;
  } else {
    data.correct = false; 
  }
}
```
BTW the data that's produced by the `image-button-response` plugin is explained in the [plugin documentation](https://www.jspsych.org/6.3/plugins/jspsych-image-button-response/#data-generated). And if you're ever unsure of what the trial data looks like, you can print it to the browser's JavaScript console in the trial's `on_finish` function like this:
```
on_finish: function(data) {
  console.log(data);
  ...
}
```
I hope this helps!

**Reply by mary-9555** [reply]
It worked ! :))
Thanks a million Becky !
